bin/flutter

```diff
@@ -16,27 +16,31 @@ set -e
 
 unset CDPATH
 
-function follow_links() {
-  cd -P "${1%/*}"
-  local file="$PWD/${1##*/}"
+# The following is enclosed in a subshell to avoid changing the cwd of the
+# caller.
+function follow_links() (
+  cd -P "$(dirname -- "$1")"
+  file="$PWD/$(basename -- "$1")"
   while [[ -h "$file" ]]; do
-    # On Mac OS, readlink -f doesn't work.
-    cd -P "${file%/*}"
-    file="$(readlink "$file")"
-    cd -P "${file%/*}"
-    file="$PWD/${file##*/}"
+    # On Mac OS, readlink -f doesn't work, so this traverses the path one link
+    # at a time, and then cds into the link destination and find out where it
+    # ends up.
+    cd -P "$(dirname -- "$file")"
+    file="$(readlink -- "$file")"
+    cd -P "$(dirname -- "$file")"
+    file="$PWD/$(basename -- "$file")"
   done
-  echo "$PWD/${file##*/}"
-}
-
-# Convert a filesystem path to a format usable by Dart's URI parser.
-function path_uri() {
-  # Reduce multiple leading slashes to a single slash.
-  echo "$1" | sed -E -e "s,^/+,/,"
-}
+  # Convert a filesystem path to a format usable by Dart's URI parser, since
+  # the Dart command line tool treats its argument as a file URI, not a
+  # filename. Multiple slashes are reduced to a single slash, since
+  # double-slashes indicate a URI "authority", and these are all supposed to
+  # be filenames. This is an edge case, since the above will only produce
+  # multiple slashes when the input resolves to the root directory.
```

Comment:
input meaning `$file`? when will that ever resolve to `/`?

When I remove the sed substitution and run this from `/`, it seems to work fine.

Reply:
If the script location is a link to `/`. Yeah, it's not even really valid.

I mean, we could just assume that it'll never happen, and not invoke `sed` at all. I have a feeling that the reason it failed in https://github.com/dart-lang/sdk/issues/28143 (the reason for that URI transform) was because of the difference between `dirname` and `${file##*/}`.

Reply:
Thanks for the link, for future excavation there's further discussion in https://github.com/flutter/flutter/issues/7279.

I agree with your assessment the substitution was probably because we used `${file##*/}`. I think we should remove the sed substitution, because if in the future we get double slashes, it will be a different regression and we should probably catch that.